docs/rfc/block-registration.md

```diff
@@ -0,0 +1,443 @@
+- `version` (`string`|`false`|`null`) - string specifying the script version number, if it has one, which is added to the URL as a query string for cache busting purposes. If the version is set to `false`, a version number is automatically added equal to current installed WordPress version. If set to `null`, no version is added. Default value: `false`.
+
+The definition is stored inside separate JSON file which ends with `.asset.json` and is located next to the JS/CSS file listed in `block.json`. WordPress will automatically detect this file through pattern matching. This option is the preferred one as we are going to provide a way to auto-generate those asset files with `@wordpress/scripts` package.
+
+**Example:**
+
+```
+build/
+├─ editor.js
+└─ editor.asset.json
+```
+
+In `block.json`:
+```json
+{ "editorScript": "build/editor.js" }
+```
+
+In `build/editor.asset.json`:
+```json
+{
+	"handle": "my-plugin-notice-editor",
+	"dependencies": [ "wp-blocks","wp-element", "wp-i18n" ],
+	"version": "3.0.0"
+}
+```
+
+## Internationalization
+
+Localized properties are automatically wrapped in `_x` function calls on the backend and the frontend of WordPress. These translations are added as an inline script to the `wp-block-library` script handle in WordPress core or to the plugin's script handle when it defines metadata definition.
+
+WordPress string discovery automatically translates these strings using the `textDomain` property specified in the `block.json` file.
+
+**Example:**
+
+```json
+{
+	"title": "My block",
+	"description": "My block is fantastic",
+	"keywords": [ "fantastic" ],
+	"textDomain": "my-plugin"
+}
+```
+
+In JavaScript, with the help of a Babel plugin, this becomes:
+
+```js
+const metadata = {
+	title: _x( 'My block', 'block title', 'my-plugin' ),
+	description: _x( 'My block is fantastic', 'block description', 'my-plugin' ),
+	keywords: [ _x( 'fantastic', 'block keywords', 'my-plugin' ) ],
+}
+```
+
+In PHP, it is transformed at runtime to code roughly equivalent to:
+
+```php
+<?php
+$metadata = array(
+	'title'       => _x( 'My block', 'block title', 'my-plugin' ),
+	'description' => _x( 'My block is fantastic!', 'block description', 'my-plugin' ),
+	'keywords'    => array( _x( 'fantastic', 'block keywords', 'my-plugin' ) ),
+);
+```
+
+Implementation should follow the existing [get_plugin_data](https://codex.wordpress.org/Function_Reference/get_plugin_data) function which parses the plugin contents to retrieve the plugin’s metadata, and it applies translations dynamically.
+
+## PHP Runtime
+
+WordPress automatically discovers all the block.json files in the plugin/core `blocks` folder and registers the corresponding block types. These block types are made available through the [block registry](https://developer.wordpress.org/reference/classes/wp_block_type_registry/) PHP class and the blocks scripts and styles are added as dependencies to the `wp-block-library` script and style handles.
```

Comment:
I didn't really notice this section 'til now. It seems a bit oddly specific a thing to commit to supporting. It's also not entirely clear in its wording: If it's discovering a file by the name of `block.json`, there could only be a maximum of one in a `blocks` folder? Or is the idea that each individual block type would be in its own sub-folder of `blocks` (which is not clarified here)? Noting also that these definitions are "block types", not "blocks", which is a small semantic point, but one which we may want to consider before committing to one or the other. Where is the "core `blocks` folder"?

Reply:
@youknowriad - do you think we need to keep it and adjust to how it works in Gutenberg core or do you agree that it's something that could be removed as it still needs to be clarified once we have some plugins using `block.json` files? As of today, we have `packages/block-library/src` folder containing `block.json` for each static block. All of them are located in their own folder next to JS and CSS files.

Reply:
I think it's very important to have this API in place and be documented in the RFC. 

I don't have a strong opinion on whether we remove from the RFC now and bring it back later or keep it and adjust as we implement it.

Reply:
IMO I almost feel like this is a WordPress-specific implementation detail, and is better suited for a separate RFC. The block schema definition is more generic and can apply to any Gutenberg environment, but how WordPress discovers and registers blocks internally seems like something that should be determined separately.

Reply:
I still think it's very related. the RFC we're writing is WordPress specific even if some parts of it are meant to be CMS agnostic like the `block.json` without the assets part.

That said, I feel this needs time to mature. I won't oppose it being removed and reincluded in follow-up PRs.

Reply:
Removed with f3456fe and moved to the follow-up issue opened: #16209.